src/main/java/org/opensearch/neuralsearch/processor/NeuralSparseTwoPhaseProcessor.java

```diff
@@ -0,0 +1,246 @@
+     */
+    @Override
+    public SearchRequest processRequest(final SearchRequest request) throws Exception {
+        if (!enabled || ratio == 0f) {
+            return request;
+        }
+        QueryBuilder queryBuilder = request.source().query();
+        // Collect the nested NeuralSparseQueryBuilder in the whole query.
+        Multimap<NeuralSparseQueryBuilder, Float> queryBuilderMap;
+        queryBuilderMap = collectNeuralSparseQueryBuilder(queryBuilder, 1.0f);
+        if (queryBuilderMap.isEmpty()) {
+            return request;
+        }
+        // Make a nestedQueryBuilder which includes all the two-phase QueryBuilder.
+        QueryBuilder nestedTwoPhaseQueryBuilder = getNestedQueryBuilderFromNeuralSparseQueryBuilderMap(queryBuilderMap);
+        nestedTwoPhaseQueryBuilder.boost(getOriginQueryWeightAfterRescore(request.source()));
+        // Add it to the rescorer.
+        RescorerBuilder<QueryRescorerBuilder> twoPhaseRescorer = buildRescoreQueryBuilderForTwoPhase(nestedTwoPhaseQueryBuilder, request);
+        request.source().addRescorer(twoPhaseRescorer);
+        return request;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    /**
+     * Based on ratio, split a Map into two map by the value.
+     * @param queryTokens the queryTokens map, key is the token String, value is the score.
+     * @param thresholdRatio The ratio that control how tokens map be split.
+     * @return A map has two element, {[True, token map whose value above threshold],[False, token map whose value below threshold]}
+     */
+    public static Map<String, SetOnce<Map<String, Float>>> splitQueryTokensByRatioedMaxScoreAsThreshold(
+        final Map<String, Float> queryTokens,
+        final float thresholdRatio
+    ) {
+
+        float max = 0f;
+        for (Float value : queryTokens.values()) {
+            max = Math.max(value, max);
+        }
+        float threshold = max * thresholdRatio;
+
+        Map<Boolean, Map<String, Float>> queryTokensByScore = queryTokens.entrySet()
+            .stream()
+            .collect(
+                Collectors.partitioningBy(entry -> entry.getValue() >= threshold, Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
+            );
+
+        SetOnce<Map<String, Float>> highScoreTokensSetOnce = new SetOnce<>(queryTokensByScore.get(Boolean.TRUE));
+        SetOnce<Map<String, Float>> lowScoreTokensSetOnce = new SetOnce<>(queryTokensByScore.get(Boolean.FALSE));
+        if (Objects.isNull(highScoreTokensSetOnce.get())) {
+            throw new IllegalArgumentException("Query tokens cannot be null.");
+        }
+        return Map.of(UP_THRESHOLD, highScoreTokensSetOnce, DOWN_THRESHOLD, lowScoreTokensSetOnce);
+    }
+
+    /**
+     * Factory to create NeuralSparseTwoPhaseProcessor, provide default parameter,
+     *
+     */
+    public static class Factory implements Processor.Factory<SearchRequestProcessor> {
```

Comment:
Can we move the Factory to the bottom of this class, and move other methods before where they are invoked?

Reply:
Moved， but other method is private, maybe should after the invoke?

Reply:
OK, I think putting the private methods after the public ones is also fine